lahja/tools/drivers/driver.py

```diff
@@ -0,0 +1,328 @@
+
+
+def serve_endpoint(config: ConnectionConfig) -> Initializer:
+    return Initializer(_serve_endpoint, config=config)
+
+
+#
+# EndpointAPI.run
+#
+def _run_endpoint(engine: EngineAPI, name: str) -> AsyncContextManager[EndpointAPI]:
+    logger.debug('[%s(%s)].run()', engine.endpoint_class.__name__, name)
+    return engine.endpoint_class(name).run()
+
+
+def run_endpoint(name) -> Initializer:
+    return Initializer(_run_endpoint, name=name)
+
+
+#
+# Serve request
+#
+GetResponseFn = Callable[[EndpointAPI, BaseRequestResponseEvent], BaseEvent]
+
+
+async def _serve_request(engine: EngineAPI,
+                         endpoint: EndpointAPI,
+                         request_type: BaseRequestResponseEvent[BaseEvent],
+                         get_response: GetResponseFn) -> None:
+    logger.debug('[%s] waiting for request: %s', endpoint, request_type)
+    request = await endpoint.wait_for(request_type)
+    logger.debug('[%s] received request: %s', endpoint, request)
+    response = get_response(endpoint, request)
+    await endpoint.broadcast(response, config=request.broadcast_config())
+    logger.debug('[%s] sent response: %s', endpoint, response)
+
+
+def serve_request(request_type: BaseRequestResponseEvent[BaseEvent],
+                  get_response: GetResponseFn) -> AsyncAction:
+    return AsyncAction(_serve_request, request_type=request_type, get_response=get_response)
+
+
+#
+# EndpointAPI.wait_for
+#
+async def _wait_for(engine: EngineAPI,
+                    endpoint: EndpointAPI,
+                    event_type: Type[BaseEvent],
+                    on_event: Optional[EventHandlerFn]) -> None:
+    logger.debug('[%s].wait_for(%s)', endpoint, event_type)
+    result = await endpoint.wait_for(event_type)
+    logger.debug('[%s].wait_for(%s) RECEIVED', endpoint, event_type)
+    if on_event is not None:
+        if asyncio.iscoroutinefunction(on_event):
+            await on_event(endpoint, result)
+        else:
+            on_event(endpoint, result)
+
+
+def wait_for(event_type: Type[BaseEvent], on_event: EventHandlerFn = None) -> AsyncAction:
+    return AsyncAction(_wait_for, event_type=event_type, on_event=on_event)
+
+
+#
+# EndpointAPI.request
+#
+async def _request(engine: EngineAPI,
+                   endpoint: EndpointAPI,
+                   event: BaseRequestResponseEvent[BaseEvent],
+                   config: Optional[BroadcastConfig],
+                   on_response: Optional[EventHandlerFn]) -> None:
+    logger.debug('[%s].request(%s, config=%s)', endpoint, event, config)
+    response = await endpoint.request(event, config=config)
+    logger.debug('[%s].request(%s) GOT RESPONSE: %s', endpoint, event, response)
+    if on_response is not None:
+        if asyncio.iscoroutinefunction(on_response):
+            await on_response(endpoint, response)
+        else:
+            on_response(endpoint, response)
+
+
+def request(event: BaseRequestResponseEvent[BaseEvent],
+            config: BroadcastConfig = None,
```

Comment:
Same here, some missing `Optional` (But I guess our strict type checking in this project will catch this anyway)